staging/src/k8s.io/api/flowcontrol/v1alpha1/types.go

```diff
@@ -0,0 +1,414 @@
+type FlowSchemaConditionType string
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// PriorityLevelConfiguration represents the configuration of a priority level.
+type PriorityLevelConfiguration struct {
+	metav1.TypeMeta `json:",inline"`
+	// `metadata` is the standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+	// `spec` is the specification of the desired behavior of a "request-priority".
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#spec-and-status
+	// +optional
+	Spec PriorityLevelConfigurationSpec `json:"spec,omitempty" protobuf:"bytes,2,opt,name=spec"`
+	// `status` is the current status of a "request-priority".
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#spec-and-status
+	// +optional
+	Status PriorityLevelConfigurationStatus `json:"status,omitempty" protobuf:"bytes,3,opt,name=status"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// PriorityLevelConfigurationList is a list of PriorityLevelConfiguration objects.
+type PriorityLevelConfigurationList struct {
+	metav1.TypeMeta `json:",inline"`
+	// `metadata` is the standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ListMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+	// `items` is a list of request-priorities.
+	Items []PriorityLevelConfiguration `json:"items" protobuf:"bytes,2,rep,name=items"`
+}
+
+// PriorityLevelConfigurationSpec is specification of a priority level
+// +union
+type PriorityLevelConfigurationSpec struct {
+	// `queuingType` indicates whether this priority level does
+	// queuing or is exempt.  Valid values are "Queuing" and "Exempt",
+	// and this field defaults to "Queuing".  "Exempt" means that
+	// requests of this priority level are not subject to concurrency
+	// limits (and thus are never queued) and do not detract from the
+	// concurrency available for non-exempt requests.
+	// Required.
+	// +unionDiscriminator
+	QueuingType PriorityLevelQueueingType `json:"queuingType" protobuf:"varint,1,opt,name=queuingType"`
+
+	// `queuing` holds the configuration parameters that are
+	// only meaningful for a priority level that does queuing (i.e.,
+	// is not exempt).  This field must be non-empty if and only if
+	// `queuingType` is `"Queuing"`.
+	// +optional
+	QueuingConfig *QueuingConfiguration `json:"queuing,omitempty" protobuf:"bytes,2,opt,name=queuing"`
+
+	// `exempt` holds the configuration parameters for exempt priority levels.
+	// +optional
+	Exempt *ExemptConfiguration `json:"exempt,omitempty" protobuf:"bytes,3,opt,name=exempt"`
```

Comment:
As it is currently written, deserialization from any JSON the user can write will put `nil` in this field.  And serialization of any possible golang value for this struct will omit this field.  I think this field should not be here.

Reply:
JSON can look like `"exempt": {}`

If it is an empty struct I agree it doesn't seem necessary at the moment since we have the discriminator.

Reply:
In fact, that is the _only_ way that the `exempt` field can be written in the JSON.  It looks spurious in the presence of the discriminator.

Reply:
There are exactly two possible JSON strings for an exempt PL config (modulo white space and field ordering):
```
{ "queuingType": "Exempt" }
```
and
```
{ "queuingType": "Exempt", "exempt": {} }
```

Both of those two mean the same thing.  In fact, they both deserailize to the same golang value: `{QueuingType: "Exempt"}`.  I see no reason to bother users with the possibility to write something more complicated that adds nothing.

Reply:
removed the definition of empty struct